### PR TITLE
Execute rule when checking permissions for guests

### PR DESCRIPTION
**Pull request by arogachev**
| Q             | A
| ------------- | ---
| Is bugfix?    | ✔️
| New feature?  | ❌
| Breaks BC?    | ❌

### CODE REVIEW

File: src/ItemsStorageInterface.php [hunk 1]
```diff
@@ -7,7 +7,7 @@
 /**
  * A storage for RBAC roles and permissions used in {@see Manager}.
  *
- * @psalm-type ItemsIndexedByName = array<string, Permission|Role>
+ * @psalm-type ItemsIndexedByName = array<string, Item>
```
Comment (vjik) [inline src/ItemsStorageInterface.php:10]:
```suggestion
 * @psalm-type ItemsIndexedByName = array<string, Permission|Role>
```

File: tests/Support/AuthorRule.php [hunk 2]
```diff
@@ -12,8 +12,8 @@
  */
 final class AuthorRule implements RuleInterface
 {
-    public function execute(string $userId, Item $item, array $parameters = []): bool
+    public function execute(?string $userId, Item $item, array $parameters = []): bool
     {
-        return $parameters['authorID'] === $userId;
+        return isset($parameters['authorID']) && $parameters['authorID'] === $userId;
```
Comment (vjik) [inline tests/Support/AuthorRule.php:17]: `isset(null) === false`, so when `$userId` is null and `$parameters['authorID']` is null this code returns false. Is it expected?

File: tests/Common/ManagerConfigurationTestTrait.php [hunk 3]
```diff
@@ -60,6 +61,8 @@ protected function createFilledManager(bool $enableDirectPermissions = false): M
             $this->assignmentsStorage,
             new SimpleRuleFactory([
                 'isAuthor' => new AuthorRule(),
+                'easyTrue' => new EasyRule(false),
```
Comment (vjik) [inline tests/Common/ManagerConfigurationTestTrait.php:64]:
```suggestion
                'easyTrue' => new EasyRule(true),
```

File: tests/Common/ManagerLogicTestTrait.php [hunk 4]
```diff
@@ -141,6 +141,22 @@ public function testUserHasPermissionWithGuest($userId, array $tests): void
...
+        $manager = $this->createFilledManager()
+            ->setGuestRoleName('guest')
+            ->addPermission(
+                (new Permission('viewIssue'))->withRuleName('easyTrue'),
+            )
+            ->addRole(new Role('guest'))
+            ->addChild('guest', 'viewIssue')
+            ->assign('guest', $userId);
+
+        $this->assertFalse($manager->userHasPermission(null, 'viewIssue'));
```
Comment (vjik) [inline tests/Common/ManagerLogicTestTrait.php:156]: With rule `easyTrue` should be true?

File: src/RuleInterface.php [hunk 5]
```diff
@@ -12,11 +12,12 @@ interface RuleInterface
...
-     * @param string $userId The user ID. This should be a string representing the unique identifier of a user.
+     * @param string|null $userId The user ID. This should be a string representing the unique identifier of a user. For
+     * guests the value is `null`.
      * @param Item $item The role or permission that this rule is associated with.
      * @param array $parameters Parameters passed to {@see CheckAccessInterface::userHasPermission()}.
      *
      * @return bool Whether the rule permits the auth item it is associated with.
      */
-    public function execute(string $userId, Item $item, array $parameters = []): bool;
+    public function execute(?string $userId, Item $item, array $parameters = []): bool;
```
Comment (vjik) [inline src/RuleInterface.php:22]: Add info about interface changes to changelog.

File: tests/Support/AuthorRule.php (hunk 2 above)
Comment (arogachev) [reply tests/Support/AuthorRule.php]: Yes.

File: CHANGELOG.md [hunk 6]
```diff
@@ -27,6 +27,8 @@
   - `userHasItem()`.
     (@arogachev)
 - Enh #165: Rename `DefaultRoleNotFoundException` to `DefaultRolesNotFoundException` and finalize it (@arogachev)
+- Bug #172: Execute rule when checking permissions for guests (@arogachev)
+- Bug #172: Make `$userId` parameter `nullable` in `RuleInterface::execute()` (@arogachev)
```
Comment (vjik) [inline CHANGELOG.md:31]:
```suggestion
- Chg #172: Make `$userId` parameter `nullable` in `RuleInterface::execute()` (@arogachev) 
```

**Pull request merged**